README.md
to run, open command prompt, cd to the root of project and type 'live-server'

ChangeLog.txt
11/4/2016
-------------------------------------------------------
	- updated documentation
	- added ability to view filter settings in the url params
	- added ability to load filter server output
	- added ability to remove layers when resetting the map
	- added default no layers view
	- resized toolbar widths to smaller size
	- fixed sidenavs so they do not overlay since filter sidenav does not have backdrop overlay and is open on initial load
	- set up methodolgy for storing layer ids for easy and accurate accessibility
	- fixed player controls so that vector layer is not accidentally counted as a layer and iterated overlaid
	- added close button for phenomenolgy and settings sidenavs
	- fixed css for help and close buttons on sidenav toolbars

10/28/2016
-------------------------------------------------------
	- added ability to copy lat and lon data to clipboard 
	- added right click control to the openlayers map
	- worked on page accessibility when viewed using smaller screen
	- added ability to create an AOI and return extents, for use on algorithm tools
	- added center option to the new right click menu on the openlayers map


10/21/2016
-------------------------------------------------------
	- added playback speed selection option to the application settings sidenav
	- redesigned layout per guidelines provided
		- 2nd toolbar moved to the bottom
		- larger play icon added
		- moved satellite widget button to the top toolbar
		- moved which side of the screen the satellite widget opens on
		- added logic so that the filter sidenav is open by default and is still overlaid on the page
		- removed excessive rotation buttons from the openlayers map
		- adjusted the 90 degree rotation button to add 90 degrees to the maps rotation no matter how many times it is clicked
		- changed the icon on the 90 degree rotation button
		- removed the full screen control from the openlayers map
		
	- adjusted appearance of filter sidenav
	- linked timeline buttons to change style depending on which image is displayed
	- added an attribution to the openlayers map tiles
	- added right click functionality to the timeline buttons.  This allows the user to select whether a tile is included in the moviemaker or not.  
		The buttons style changes based on whether it is selected or not for the moviemaker.

10/14/2016
-------------------------------------------------------
	- added movie icon to the products controls area on the map toolbar
	- added photo icon to the products controls area on the map toolbar 
	- cleaned up timeline and linked to the map imagery
	- added logic for the play controls

10/7/2016
-------------------------------------------------------
	- added satellite location icon to open angular toast
	- added angular material toast element linked to satellite icon
	- embedded d3 based rotating earth in toast element
	- corrected positioning of elements in the map controls toolbar
	- fixed intro.js bubble positioning bug occurring for help tutorials on right hand sidenavs
	- added timeline capability to the controls toolbar
	- linked image button in timeline to change displayed image on map

9/30/2016
-------------------------------------------------------
	- added help tutorial to the algorithm sidenav
	- added help tutorial to the filter sidenav
	- changed help tutorial code so that angular-intro could be used for each tutorial

9/23/2016
-------------------------------------------------------
	- added ability to maintain map rotation when playing forward or backward through image stack
	- added ability to maintain map location when playing forward or backward through image stack
	- added ability to maintain zoom level when playing forward or backward through image stack
	- add default rotation buttons to map in MapService.js
	- move location of 0 degree orientation button from right upper corner to under the zoom controls
	- add icons for movie and image and add to map container toolbar
	- add additional help bubbles to the app tutorial
	- make help bubble text justified
	- fix map not loading on initial page load issue 
	- added images and info to the info sidenav
	- changed which side of screen the info sidenav opens from
	- changed which side of screen the phenomenology sidenav opens from
	- moved info button from top level toolbar to the help options sidenav
	- moved filter button in top level toolbar from left side to right side
	- fixed custom openlayers build error after adding default rotation buttons to the openlayers map
	

9/16/2016
-------------------------------------------------------
	- added logic to link player controls to the map image
	- add beginning, end, forward, backward, play, and stop buttons to the map container toolbar
	- add fullscreen functionality
	- remove search bar and search results sidenav
	- moved algorithm and filter buttons from map container toolbar to top level toolbar on left side
	- moved openlayers code from index.js to MapService.js
	- added methods in MapService.js to reset the map area 
	- fixed custom openlayers build error after adding fullscreen capability to the openlayers map

9/9/2016
-------------------------------------------------------
	- added intro.js functionality to display help bubbles
	- added links in algorithm sidenav to provided list of algorithms options dialog
	- added options dialog for each alogorithm

9/2/2016
-------------------------------------------------------
	- added custom openlayers build 

8/26/2016
-------------------------------------------------------
	- added openlayers functionality to the app in index.js

8/19/2016
-------------------------------------------------------
	- added top level toolbar
	- added map container toolbar
	- added algorithm sidenav
	- added filter option sidenav
	- added help options sidenav
	- added algorithm, help, filter buttons
	- downloaded and imported multiple icons
	- started with angularjs material tutorial app and cut out unnecessary code



app/src/scrollbox.js
 /*global jQuery */
/*!
 * jQuery Scrollbox
 * (c) 2009-2013 Hunter Wu <hunter.wu@gmail.com>
 * MIT Licensed.
 *
 * http://github.com/wmh/jquery-scrollbox
 */

(function($) {

$.fn.scrollbox = function(config) {
  //default config
  var defConfig = {
    linear: false,          // Scroll method
    startDelay: 2,          // Start delay (in seconds)
    delay: 3,               // Delay after each scroll event (in seconds)
    step: 5,                // Distance of each single step (in pixels)
    speed: 32,              // Delay after each single step (in milliseconds)
    switchItems: 1,         // Items to switch after each scroll event
    direction: 'vertical',
    distance: 'auto',
    autoPlay: true,
    onMouseOverPause: true,
    paused: false,
    queue: null,
    listElement: 'md-list',
    listItemElement:'md-list-item',
    infiniteLoop: true,     // Infinite loop or not
    switchAmount: 0,        // Give a number if you don't want to have infinite loop
    afterForward: null,     // Callback function after each forward action
    afterBackward: null,    // Callback function after each backward action
    triggerStackable: false // Allow triggers when action is not finish yet
  };
  config = $.extend(defConfig, config);
  config.scrollOffset = config.direction === 'vertical' ? 'scrollTop' : 'scrollLeft';
  if (config.queue) {
    config.queue = $('#' + config.queue);
  }

  return this.each(function() {
    var container = $(this),
        containerUL,
        scrollingId = null,
        nextScrollId = null,
        paused = false,
        releaseStack,
        backward,
        forward,
        resetClock,
        scrollForward,
        scrollBackward,
        forwardHover,
        pauseHover,
        switchCount = 0,
        stackedTriggerIndex = 0;

    if (config.onMouseOverPause) {
      container.bind('mouseover', function() { paused = true; });
      container.bind('mouseout', function() { paused = false; });
    }
    containerUL = container.children(config.listElement + ':first-child');

    // init default switchAmount
    if (config.infiniteLoop === false && config.switchAmount === 0) {
      config.switchAmount = containerUL.children().length;
    }

    scrollForward = function() {
      if (paused) {
        return;
      }
      var curLi,
          i,
          newScrollOffset,
          scrollDistance,
          theStep;

      curLi = containerUL.children(config.listItemElement + ':first-child');

      scrollDistance = config.distance !== 'auto' ? config.distance :
        config.direction === 'vertical' ? curLi.outerHeight(true) : curLi.outerWidth(true);

      // offset
      if (!config.linear) {
        theStep = Math.max(3, parseInt((scrollDistance - container[0][config.scrollOffset]) * 0.3, 10));
        newScrollOffset = Math.min(container[0][config.scrollOffset] + theStep, scrollDistance);
      } else {
        newScrollOffset = Math.min(container[0][config.scrollOffset] + config.step, scrollDistance);
      }
      container[0][config.scrollOffset] = newScrollOffset;

      if (newScrollOffset >= scrollDistance) {
        for (i = 0; i < config.switchItems; i++) {
          if (config.queue && config.queue.find(config.listItemElement).length > 0) {
            containerUL.append(config.queue.find(config.listItemElement)[0]);
            containerUL.children(config.listItemElement + ':first-child').remove();
          } else {
            containerUL.append(containerUL.children(config.listItemElement + ':first-child'));
          }
          ++switchCount;
        }
        container[0][config.scrollOffset] = 0;
        clearInterval(scrollingId);
        scrollingId = null;

        if ($.isFunction(config.afterForward)) {
          config.afterForward.call(container, {
            switchCount: switchCount,
            currentFirstChild: containerUL.children(config.listItemElement + ':first-child')
          });
        }
        if (config.triggerStackable && stackedTriggerIndex !== 0) {
          releaseStack();
          return;
        }
        if (config.infiniteLoop === false && switchCount >= config.switchAmount) {
          return;
        }
        if (config.autoPlay) {
          nextScrollId = setTimeout(forward, config.delay * 1000);
        }
      }
    };

    // Backward
    // 1. If forwarding, then reverse
    // 2. If stoping, then backward once
    scrollBackward = function() {
      if (paused) {
        return;
      }
      var curLi,
          i,
          newScrollOffset,
          scrollDistance,
          theStep;

      // init
      if (container[0][config.scrollOffset] === 0) {
        for (i = 0; i < config.switchItems; i++) {
          containerUL.children(config.listItemElement + ':last-child').insertBefore(containerUL.children(config.listItemElement+':first-child'));
        }

        curLi = containerUL.children(config.listItemElement + ':first-child');
        scrollDistance = config.distance !== 'auto' ?
            config.distance :
            config.direction === 'vertical' ? curLi.height() : curLi.width();
        container[0][config.scrollOffset] = scrollDistance;
      }

      // new offset
      if (!config.linear) {
        theStep = Math.max(3, parseInt(container[0][config.scrollOffset] * 0.3, 10));
        newScrollOffset = Math.max(container[0][config.scrollOffset] - theStep, 0);
      } else {
        newScrollOffset = Math.max(container[0][config.scrollOffset] - config.step, 0);
      }
      container[0][config.scrollOffset] = newScrollOffset;

      if (newScrollOffset === 0) {
        --switchCount;
        clearInterval(scrollingId);
        scrollingId = null;

        if ($.isFunction(config.afterBackward)) {
          config.afterBackward.call(container, {
            switchCount: switchCount,
            currentFirstChild: containerUL.children(config.listItemElement + ':first-child')
          });
        }
        if (config.triggerStackable && stackedTriggerIndex !== 0) {
          releaseStack();
          return;
        }
        if (config.autoPlay) {
          nextScrollId = setTimeout(forward, config.delay * 1000);
        }
      }
    };

    releaseStack = function () {
      if (stackedTriggerIndex === 0) {
        return;
      }
      if (stackedTriggerIndex > 0) {
        stackedTriggerIndex--;
        nextScrollId = setTimeout(forward, 0);
      } else {
        stackedTriggerIndex++;
        nextScrollId = setTimeout(backward, 0);
      }
    };

    forward = function() {
      clearInterval(scrollingId);
      scrollingId = setInterval(scrollForward, config.speed);
    };

    backward = function() {
      clearInterval(scrollingId);
      scrollingId = setInterval(scrollBackward, config.speed);
    };

    // Implements mouseover function.
    forwardHover = function() {
        config.autoPlay = true;
        paused = false;
        clearInterval(scrollingId);
        scrollingId = setInterval(scrollForward, config.speed);
    };
    pauseHover = function() {
        paused = true;
    };

    resetClock = function(delay) {
      config.delay = delay || config.delay;
      clearTimeout(nextScrollId);
      if (config.autoPlay) {
        nextScrollId = setTimeout(forward, config.delay * 1000);
      }
    };

    if (config.autoPlay) {
      nextScrollId = setTimeout(forward, config.startDelay * 1000);
    }

    // bind events for container
    container.bind('resetClock', function(delay) { resetClock(delay); });
    container.bind('forward', function() {
      if (config.triggerStackable) {
        if (scrollingId !== null) {
          stackedTriggerIndex++;
        } else {
          forward();
        }
      } else {
        clearTimeout(nextScrollId);
        forward();
      }
    });
    container.bind('backward', function() {
      if (config.triggerStackable) {
        if (scrollingId !== null) {
          stackedTriggerIndex--;
        } else {
          backward();
        }
      } else {
        clearTimeout(nextScrollId);
        backward();
      }
    });
    container.bind('pauseHover', function() { pauseHover(); });
    container.bind('forwardHover', function() { forwardHover(); });
    container.bind('speedUp', function(event, speed) {
      if (speed === 'undefined') {
        speed = Math.max(1, parseInt(config.speed / 2, 10));
      }
      config.speed = speed;
    });

    container.bind('speedDown', function(event, speed) {
      if (speed === 'undefined') {
        speed = config.speed * 2;
      }
      config.speed = speed;
    });

    container.bind('updateConfig', function (event, options) {
        config = $.extend(config, options);
    });

  });
};

}(jQuery));



app/src/Main/MapService.js
angular
    .module('main')
    .service('MapService', function($location, DefaultConfigs) {

        this.map = null;
        this.layers = [];
        this.proj = null;
        this.baseURL = 'http://127.0.0.1:8888/';
        this.currentLayers = [];
        this.zoomLevel = DefaultConfigs.optionDefaults.zoom;
        this.currentZoomLevel = this.zoomLevel;
        this.centerValue = [DefaultConfigs.optionDefaults.x, DefaultConfigs.optionDefaults.y];
        this.currentCenterValue = this.centerValue;
        this.currentRotationValue = 0;
        this.tileAttribution = new ol.Attribution({ html: '<img src="./assets/images/af.png"> <img src="./assets/images/nasic.png">'});

        function l(id, playerId) {
            this.id = id;
            this.playerId = playerId;
        }

        function copyToClipboard(data) {
            angular.element('<textarea/>')
                .css({ 'opacity' : '0', 'position' : 'fixed' })
                .text(data)
                .appendTo(angular.element(window.document.body))
                .select()
                .each(function() { document.execCommand('copy') })
                .remove();
        };

        // creates a map projection
        this.createProjection = function() {
            this.proj = new ol.proj.Projection({
                code: 'EPSG:0000',
                units: 'pixels',
                extent: [0, 0, 8192, 8192]
            });
        };

        // creates a map layer
        this.createLayer = function(ind, isVisible=false, counter=1) {
            imageLayer = new ol.layer.Tile({
                preload: Infinity,
                source: new ol.source.XYZ({
                    url: this.baseURL + ind + '/{z}-{y}-{x}.png',
                    projection: this.proj,
                    wrapX: false,
                    attributions: [this.tileAttribution]
                }),
                transitionEffect: 'resize',
                visible: isVisible
            });
            imageLayer.id = counter;
            imageLayer.playerId = parseInt(ind, 10);

            return imageLayer;
        };

        // updates the map layers
        this.buildLayers = function() {
            var urlSettings = $location.search();
            var temp = null;
            var counter = -1;
            this.currentLayers = [];
            
            for (i = parseInt(urlSettings.start, 10); i < (parseInt(urlSettings.end, 10)+1); i++) {
                console.log('i is: ' + i);
                counter = counter + 1;
                var temp = new l(counter, i);
                this.currentLayers.push(temp);
                temp = this.createLayer(ind=i, isVisible=false, counter=counter);
                this.layers.push(temp);
            }

            vectorLayer = new ol.layer.Vector({
                source: new ol.source.Vector()
            });

            this.layers.push(vectorLayer);

        };

        // 90° rotation controls
        this.RotateCW90 = function(opt_options) {
            var options = opt_options || {};
            var button = document.createElement('button');
            button.innerHTML = '<img src="./assets/svg/rotate_90_18px.svg" alt="Rotate 90" >';
            var this_ = this;

            var handleRotate90 = function() {
                var temp = this_.getMap().getView().getRotation()
                this_.getMap().getView().setRotation(temp + Math.PI/2);
            };

            button.addEventListener('click', handleRotate90, false);
            button.addEventListener('touchstart', handleRotate90, false);

            var element = document.createElement('div');
            element.className = 'rotate-90 ol-unselectable ol-control';
            element.appendChild(button);

            ol.control.Control.call(this, {
                element: element,
                target: options.target
            });
        };

        ol.inherits(this.RotateCW90, ol.control.Control);

        // creates an openlayers map
        this.createMap = function(target) {
            this.createProjection();
            this.buildLayers();

            if (this.currentLayers.length != 0) {
                // create control for adding mouse coordinates to the map
                var mousePositionControl = new ol.control.MousePosition({
                    coordinateFormat: ol.coordinate.createStringXY(4),
                    projection: DefaultConfigs.getProjection(),
                    undefinedHTML: '&nbsp;'
                });

                // create interaction for adding dragBox to the map
                var dragBox = new ol.interaction.DragBox({
                    condition: ol.events.condition.platformModifierKeyOnly
                });
                // listener for dragBox to acquire the dimensions
                dragBox.on('boxend', function() {
                    var extent = dragBox.getGeometry().getExtent();
                    console.log(extent);
                });
                // listener incase action is needed at end of box creation
                dragBox.on('boxstart', function() {
                });

                // create interaction for adding extent to the map
                this.extentBox = new ol.interaction.Extent({
                    condition: ol.events.condition.altKeyOnly
                });
                this.extentBox.setActive(false);

                // create the map 
                this.map = new ol.Map({
                    layers: this.layers,
                    target: target,
                    renderer: 'canvas',
                    controls: ol.control.defaults().extend([
                        //new ol.control.FullScreen(),
                        //this.contextmenu,
                        new this.RotateCW90(),
                        mousePositionControl
                    ]),
                    interactions: ol.interaction.defaults().extend([
                        dragBox,
                        this.extentBox
                    ]),
                    view: new ol.View({
                        center: this.currentCenterValue,
                        zoom: this.currentZoomLevel,
                        maxZoom: 5,
                        projection: DefaultConfigs.getProjection(),
                        rotation: this.currentRotationValue
                    })
                });

                var url_clipboard = './assets/svg/clipboard_48px.svg';
                var url_marker = './assets/svg/location_on_48px.svg';
                var url_unmarker = './assets/svg/location_off_48px.svg';
                var url_center = './assets/svg/searching_48px.svg';

                // from https://github.com/DmitryBaranovskiy/raphael
                var elastic = function(t) {
                    return Math.pow(2, -10 * t) * Math.sin((t - 0.075) * (2 * Math.PI) / 0.3) + 1;
                };

                var center = function(obj, foo = this.map){
                    var pan = ol.animation.pan({
                        duration: 1000,
                        easing: elastic,
                        source: foo.getView().getCenter()
                    });

                    foo.beforeRender(pan);
                    foo.getView().setCenter(obj.coordinate);
                };

                var copycoord = function(event) {
                    //var coord = ol.proj.transform(event.coordinate, 'EPSG:0000', 'EPSG:4326');
                    var lon = event.coordinate[0];
                    var lat = event.coordinate[1];
                    console.log('The latitude is: ' + lon + ' and the longitude is ' + lat);
                    copyToClipboard(lat + ', ' + lon);
                    //alert('The latitude is: ' + lon + ' and the longitude is ' + lat);
                };

                this.contextmenu_items = [
                    {
                        text: 'Center map here',
                        callback: center,
                        icon: url_center
                    },
                    //'-', this is a separator
                    {
                        text: 'Copy Coordinates',
                        callback: copycoord,
                        icon: url_clipboard
                    }
                ];

                this.contextmenu = new ContextMenu({
                    width: 190,
                    default_items: false,
                    items: this.contextmenu_items
                });

                this.map.addControl(this.contextmenu);
            }
        };

        // removes a map
        this.resetMap = function() {
            // Remove the old map if there is one
            if (typeof this.map !== 'undefined' && this.map !== null) {
                $('.ol-viewport').remove();
            }
            this.map = null;
            this.layers = [];
            console.log('The map was reset');
        };

        // retain current zoom level
        this.updateCurrentZoom = function() {
            this.currentZoomLevel = this.map.getView().getZoom();
        };

        // retain current view center
        this.updateCurrentCenter = function() {
            this.currentCenterValue = this.map.getView().getCenter();
        };

        // retain current view rotation
        this.updateCurrentRotation = function() {
            this.currentRotationValue = this.map.getView().getRotation();
        };

        // update layers on map
        this.updateLayers = function(layers) {
            this.map.getLayers().insertAt(0, layers);
        };

});

app/src/Main/DefaultConfigs.js
angular
    .module('main')
    
    .factory('DefaultConfigs', function() {
        currentDate = new Date(Date.now());
        currentTime = (currentDate.getHours()<10?'0':'') + currentDate.getHours() + ':' + (currentDate.getMinutes()<10?'0':'') + currentDate.getMinutes();

        apps = [{ 'id': '1',
                  'name': 'Spectre',
                  'isDisabled': true
                }, {
                  'id': '2',
                  'name': 'Skyfall',
                  'isDisabled': false
                }];

        return {
            'selectedApp': apps[1],
            'classificationText': 'UNCLASSIFIED',
            'classificationClass': 'classif-unclass',
            'defaultSatelliteOption': '3',
            'liveFeedSelected': true,
            'defaultDate': currentDate,
            'defaultTime': currentTime,
            'defaultDelta': '',
            'defaultAOI': '',

            'tiles': {
                '1': {
                    'prefix': 'http://api.tiles.mapbox.com/v4/',
                    'suffix': '/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoiYm93bWFubWMiLCJhIjoieE9WenlhayJ9.QFS8jQtCusMhwwVSMQIg9w'
                },
                '2': {
                    'prefix': 'http://api.tiles.mapbox.com/v4/',
                    'suffix': '/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoiYm93bWFubWMiLCJhIjoieE9WenlhayJ9.QFS8jQtCusMhwwVSMQIg9w'
                },
                'heo1': {
                    'prefix': 'http://api.tiles.mapbox.com/v4/',
                    'suffix': '/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoiYm93bWFubWMiLCJhIjoieE9WenlhayJ9.QFS8jQtCusMhwwVSMQIg9w'
                }
            },

            'getProjection': function() {
                return new ol.proj.Projection({
                    //code: 'EPSG:3857',
                    code: 'EPSG:0000',
                    units: 'pixels',
                    extent: [0, 0, 8192, 8192]
                });
            },

            'getCoordinateTransforms': function(sensorId) {
                return null;
            },

            // how long to leave each "frame" on the screen in ms
            'updateInterval': 750,

            'optionDefaults': {
                'aoi': '3',
                'sensor': '1',
                'delta': '2',     // 2 hours back
                'end': null,
                'livefeed': true,
                'x': 4096,
                'y': 4096,
                'zoom': 2,
                'idx': 0
            },

            'availableSensors': [{
                'id': '1',
                'name': 'GEO1',
                'isDisabled': false
            }, {
                'id': '2',
                'name': 'GEO2',
                'isDisabled': false
            },{
                'id': '3',
                'name': 'HEO33',
                'isDisabled': false
            },{
                'id': '4',
                'name': 'HEO34',
                'isDisabled': false
            }],

            'availableAois': [{
                'id': '3',
                'name': '3 - Ohio',
                'isDisabled': false
            }, {
                'id': '4',
                'name': '4 - Florida',
                'isDisabled': false
            }, {
                'id': '5',
                'name': '5 - North Carolina',
                'isDisabled': false
            }],

            'validDeltas': [{
                'id': '2',
                'name': '2 Hours',
                'isDisabled': false
            }, {
                'id': '4',
                'name': '4 Hours',
                'isDisabled': false
            }, {
                'id': '6',
                'name': '6 Hours',
                'isDisabled': false
            }, {
                'id': '12',
                'name': '12 Hours',
                'isDisabled': false
            }, {
                'id': '24',
                'name': '24 Hours',
                'isDisabled': false
            }, {
                'id': '48',
                'name': '48 Hours',
                'isDisabled': false
            }, {
                'id': '72',
                'name': '72 Hours',
                'isDisabled': false
            }, {
                'id': '576',
                'name': '24 Days',
                'isDisabled': false
            }]
        };

    });


app/assets/images/CORSHTTPRequestHandler.py
import SimpleHTTPServer
class CORSHTTPRequestHandler(SimpleHTTPServer.SimpleHTTPRequestHandler):
    def send_head(self):
        """Common code for GET and HEAD commands.

        This sends the response code and MIME headers.

        Return value is either a file object (which has to be copied
        to the outputfile by the caller unless the command was HEAD,
        and must be closed by the caller under all circumstances), or
        None, in which case the caller has nothing further to do.

        """
        path = self.translate_path(self.path)
        f = None
        if os.path.isdir(path):
            if not self.path.endswith('/'):
                # redirect browser - doing basically what apache does
                self.send_response(301)
                self.send_header("Location", self.path + "/")
                self.end_headers()
                return None
            for index in "index.html", "index.htm":
                index = os.path.join(path, index)
                if os.path.exists(index):
                    path = index
                    break
            else:
                return self.list_directory(path)
        ctype = self.guess_type(path)
        try:
            # Always read in binary mode. Opening files in text mode may cause
            # newline translations, making the actual size of the content
            # transmitted *less* than the content-length!
            f = open(path, 'rb')
        except IOError:
            self.send_error(404, "File not found")
            return None
        self.send_response(200)
        self.send_header("Content-type", ctype)
        fs = os.fstat(f.fileno())
        self.send_header("Content-Length", str(fs[6]))
        self.send_header("Last-Modified", self.date_time_string(fs.st_mtime))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        return f


if __name__ == "__main__":
    import os
    import SocketServer

    PORT = 31338

    Handler = CORSHTTPRequestHandler
    #Handler = SimpleHTTPServer.SimpleHTTPRequestHandler

    httpd = SocketServer.TCPServer(("", PORT), Handler)

    print "serving at port", PORT
    httpd.serve_forever()




app/src/index.js
angular
    .module('SpectreSkyfall', ['ngMaterial', 'angular-intro', 'main'])
    .config(function($mdThemingProvider, $mdIconProvider){

        $mdIconProvider
            .defaultIconSet("./assets/svg/avatars.svg" , 128)
            .icon("algo"       , "./assets/svg/algorithm_48px.svg" , 48)
            .icon("clipboard"  , "./assets/svg/clipboard_48px.svg" , 48)
            .icon("close"      , "./assets/svg/close_48px.svg" , 48)
            .icon("earth"      , "./assets/svg/earth_48px.svg", 48)
            .icon("earth2"     , "./assets/svg/earth_2_48px.svg", 48)
            .icon("blueeye"    , "./assets/svg/blueeye_48px.svg", 48)
            .icon("bluerect"   , "./assets/svg/tile_blue.svg", 48)
            .icon("photo"      , "./assets/svg/photo_48px.svg" , 48)
            .icon("filter"     , "./assets/svg/filter_48px.svg" , 48)
            .icon("firstpagearrow" , "./assets/svg/first_page_arrow_48px.svg" , 48)
            .icon("help"       , "./assets/svg/help_48px.svg" , 48)
            .icon("info"       , "./assets/svg/info_empty_48px.svg" , 48)
            .icon("lastpagearrow" , "./assets/svg/last_page_arrow_48px.svg" , 48)
            .icon("leftarrow"  , "./assets/svg/left_arrow_48px.svg" , 48)
            .icon("listmenu"   , "./assets/svg/list_48px.svg" , 48)
            .icon("livehelp"   , "./assets/svg/live_help_48px.svg" , 48)
            .icon("locationon" , "./assets/svg/location_on_48px.svg" , 48)
            .icon("locationoff"  , "./assets/svg/location_off_48px.svg" , 48)
            .icon("menu"       , "./assets/svg/menu.svg" , 24)
            .icon("movie"      , "./assets/svg/movies_48px.svg" , 48)
            .icon("playarrow"  , "./assets/svg/play3.svg" , 48)
            .icon("redeye"     , "./assets/svg/redeye_48px.svg", 48)
            .icon("redrect"    , "./assets/svg/tile_red.svg", 48)
            .icon("rightarrow" , "./assets/svg/right_arrow_48px.svg" , 48)
            .icon("rotate"     , "./assets/svg/rotate_90_18px.svg" , 48)
            .icon("satellite"  , "./assets/svg/satellite_48px.svg" , 48)
            .icon("searching"  , "./assets/svg/searching_48px.svg" , 48)
            .icon("settings"   , "./assets/svg/settings_48px.svg" , 48)
            .icon("share"      , "./assets/svg/share.svg" , 24)
            .icon("stop"       , "./assets/svg/stop_48px.svg" , 48);
            

            $mdThemingProvider.theme('default')
                .primaryPalette('blue')
                .dark();

    });
